File: Dummy/hawkshaw/library/preferences/PreferenceProperty.java
package com.hawkshaw.library.preferences;

import Y1.K;
import android.content.SharedPreferences;
import i3.q;
import l3.C0709a;
import p3.C0867h;

public final class PreferenceProperty<T> implements C0709a {
    private final T defaultValue;
    private final q getter;
    private final String key;
    private final q setter;

    public PreferenceProperty(String str, T t4, q qVar, q qVar2) {
        K.n(str, "key");
        K.n(qVar, "getter");
        K.n(qVar2, "setter");
        this.key = str;
        this.defaultValue = t4;
        this.getter = qVar;
        this.setter = qVar2;
    }

    public T getValue(Object obj, C0867h hVar) {
        K.n(obj, "thisRef");
        K.n(hVar, "property");
        return this.getter.d(Prefs.INSTANCE.getSharedPrefs(), this.key, this.defaultValue);
    }

    public void setValue(Object obj, C0867h hVar, T t4) {
        K.n(obj, "thisRef");
        K.n(hVar, "property");
        if (t4 == null) {
            Prefs.INSTANCE.remove(this.key);
            return;
        }
        q qVar = this.setter;
        SharedPreferences.Editor edit = Prefs.INSTANCE.getSharedPrefs().edit();
        K.m(edit, "edit(...)");
        ((SharedPreferences.Editor) qVar.d(edit, this.key, t4)).apply();
    }
}

File: app/src/main/java/com/hawkshaw/library/preferences/PreferenceProperty.kt
package com.hawkshaw.library.preferences

import android.content.SharedPreferences
import kotlin.properties.ReadWriteProperty
import kotlin.reflect.KProperty
import android.util.Log // Added for logging
import androidx.core.content.edit

/**
 * A property delegate for SharedPreferences that handles getting and setting values.
 *
 * @param key The key to use in SharedPreferences
 * @param defaultValue The default value to return if the key is not found
 * @param getter Function to get a value from SharedPreferences
 * @param setter Function to set a value in SharedPreferences
 */
class PreferenceProperty<T>(
    private val key: String,
    private val defaultValue: T?,
    private val getter: (SharedPreferences, String, T?) -> T?,
    private val setter: (SharedPreferences.Editor, String, T?) -> SharedPreferences.Editor
) : ReadWriteProperty<Any, T?> {

    private companion object { // Companion object to hold the TAG
        private const val TAG = "PreferenceProperty"
    }

    // It might be useful to log when an instance is created,
    // but typically key/defaultValue are logged by the factory in ExtKt.kt.
    // init {
    //     Log.d(TAG, "[DEBUG] Instance created for Key: \"$key\", DefaultValue: $defaultValue")
    // }

    /**
     * Gets the value from SharedPreferences
     */
    override fun getValue(thisRef: Any, property: KProperty<*>): T? {
        Log.d(TAG, "[DEBUG] getValue called. Ref: $thisRef, Property: ${property.name}, Key: \"$key\", DefaultValue: $defaultValue")
        val value = getter(Prefs.sharedPrefs, key, defaultValue)
        Log.d(TAG, "[DEBUG] getValue for Key: \"$key\" returned value: $value")
        return value
    }

    /**
     * Sets the value in SharedPreferences
     */
    override fun setValue(thisRef: Any, property: KProperty<*>, value: T?) {
        Log.d(TAG, "[DEBUG] setValue called. Ref: $thisRef, Property: ${property.name}, Key: \"$key\", NewValue: $value")
        if (value == null) {
            Log.d(TAG, "[DEBUG] setValue for Key: \"$key\": NewValue is null. Calling Prefs.remove(\"$key\").")
            Prefs.remove(key)
            Log.d(TAG, "[DEBUG] setValue for Key: \"$key\": Prefs.remove(\"$key\") completed.")
            return
        }
        
        Log.d(TAG, "[DEBUG] setValue for Key: \"$key\": NewValue is not null. Getting SharedPreferences.Editor.")
        Prefs.sharedPrefs.edit {
            Log.d(TAG, "[DEBUG] setValue for Key: \"$key\": Invoking setter lambda.")
            setter(this, key, value) // The setter lambda itself returns the editor
            Log.d(
                TAG,
                "[DEBUG] setValue for Key: \"$key\": Setter lambda invoked. Calling apply() on editor."
            )
        }
        Log.d(TAG, "[DEBUG] setValue for Key: \"$key\": editor.apply() completed.")
    }
}
